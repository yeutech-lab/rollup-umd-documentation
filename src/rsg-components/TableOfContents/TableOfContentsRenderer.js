import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import cn from 'classnames';
import omit from 'lodash.omit';
import mapToCssModules from 'map-to-css-modules/lib';
import Nav from 'bootstrap-styled/lib/Nav';
import Form from 'bootstrap-styled/lib/Form';
import Input from 'bootstrap-styled/lib/Input';


export const defaultProps = {
  theme: {
    styleguide: {
      '$rsg-toc-display': 'block',
      '$rsg-toc-padding': '0',
      '$rsg-toc-form-width': '60%',
      '$rsg-toc-form-margin': '0 auto',
    },
  },
};

export const propTypes = {
  className: PropTypes.string, // eslint-disable-line react/require-default-props
  children: PropTypes.node, // eslint-disable-line react/require-default-props
  searchTerm: PropTypes.string.isRequired,
  onSearchTermChange: PropTypes.func.isRequired,
  /** Theme variables. Can be: */
  theme: PropTypes.shape({
    styleguide: PropTypes.shape({
      '$rsg-toc-display': PropTypes.string,
      '$rsg-toc-padding': PropTypes.string,
      '$rsg-toc-form-width': PropTypes.string,
      '$rsg-toc-form-margin': PropTypes.string,
    }),
  }),
  /**
   * Replace or remove a className from the component.
   * See example <a href="https://www.npmjs.com/package/map-to-css-modules" target="_blank">here</a>.
   */
  cssModule: PropTypes.object, // eslint-disable-line react/require-default-props
};

const TableOfContentsRendererUnstyled = (props) => {
  const {
    className,
    children,
    searchTerm,
    onSearchTermChange,
    cssModule,
    ...attributes
  } = omit(props, ['theme']);
  return (
    <div
      className={mapToCssModules(cn(className, 'rsg-toc'), cssModule)}
      {...attributes}
    >
      <Form className="rsg-toc-form">
        <Input
          className="rsg-toc-form-input"
          size="sm"
          value={searchTerm}
          placeholder="Filter by name"
          aria-label="Filter by name"
          onChange={(event) => onSearchTermChange(event.target.value)}
        />
      </Form>
      <Nav className="rsg-toc-nav">
        {children}
      </Nav>
    </div>
  );
};

TableOfContentsRendererUnstyled.defaultProps = defaultProps;
TableOfContentsRendererUnstyled.propTypes = propTypes;

const TableOfContentsRenderer = styled(TableOfContentsRendererUnstyled)`
  ${(props) => `
    &.rsg-toc {
      .rsg-toc-form {
        background: #B31255;
        padding: 0 0 20px 0;
        .rsg-toc-form-input {
          width: ${props.theme.styleguide['$rsg-toc-form-width']};
          margin: ${props.theme.styleguide['$rsg-toc-form-margin']};
        }
      }
      .rsg-toc-nav {
        padding: 20px 0 0 0 !important;
        background: white;
        text-align: left;
        display: ${props.theme.styleguide['$rsg-toc-display']};
        padding: ${props.theme.styleguide['$rsg-toc-padding']};
      }
    }
  `}
`;

TableOfContentsRenderer.defaultProps = defaultProps;
TableOfContentsRenderer.propTypes = propTypes;

export default TableOfContentsRenderer;
