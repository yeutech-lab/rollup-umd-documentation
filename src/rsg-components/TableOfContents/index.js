export { default } from '../TableOfContents/TableOfContents';
