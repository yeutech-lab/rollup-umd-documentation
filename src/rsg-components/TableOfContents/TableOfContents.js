import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ComponentsList from 'rsg-components/ComponentsList';
import TableOfContentsRenderer from 'rsg-components/TableOfContents/TableOfContentsRenderer';
import filterSectionsByName from '../../utils/filterSectionsByName';

export default class TableOfContents extends Component {
	static propTypes = {
		sections: PropTypes.array.isRequired,
		useIsolatedLinks: PropTypes.bool,
	};
	state = {
		searchTerm: '',
	};

	renderLevel(sections, useIsolatedLinks = false) {
		const items = sections.map(section => {
			const children = [...(section.sections || []), ...(section.components || [])];
			return Object.assign({}, section, {
				heading: !!section.name && children.length > 0,
				content: children.length > 0 && this.renderLevel(children, useIsolatedLinks),
			});
		});
		return <ComponentsList items={items} useIsolatedLinks={useIsolatedLinks} />;
	}

	renderSections() {
		const { searchTerm } = this.state;
		const { sections, useIsolatedLinks } = this.props;
		// If there is only one section, we treat it as a root section
		// In this case the name of the section won't be rendered and it won't get left padding
		const firstLevel = sections.length === 1 ? sections[0].components : sections;
		const filtered = filterSectionsByName(firstLevel, searchTerm);

		return this.renderLevel(filtered, useIsolatedLinks);
	}

	render() {
		const { searchTerm } = this.state;
		return (
			<TableOfContentsRenderer
				searchTerm={searchTerm}
				onSearchTermChange={searchTerm => this.setState({ searchTerm })}
			>
				{this.renderSections()}
			</TableOfContentsRenderer>
		);
	}
}
