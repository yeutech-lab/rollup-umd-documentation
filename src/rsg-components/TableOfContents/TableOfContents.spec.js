import React from 'react';
import noop from 'lodash/noop';
import TableOfContents from './TableOfContents';
import { TableOfContentsRenderer } from './TableOfContentsRenderer';

const components = [
	{
		name: 'Button',
		slug: 'button',
	},
	{
		name: 'Input',
		slug: 'input',
	},
	{
		name: 'Textarea',
		slug: 'textarea',
	},
];

const sections = [
	{
		name: 'Introduction',
		slug: 'introduction',
		content: 'intro.md',
	},
	{
		name: 'Buttons',
		slug: 'buttons',
		components: [
			{
				name: 'Button',
				slug: 'button',
			},
		],
	},
	{
		name: 'Forms',
		slug: 'forms',
		components: [
			{
				name: 'Input',
				slug: 'input',
			},
			{
				name: 'Textarea',
				slug: 'textarea',
			},
		],
	},
];

it('should render a renderer', () => {
	const actual = shallow(<TableOfContents sections={[{ components }]} />);

	expect(actual).toMatchSnapshot();
});

it('should render renderer with sections with nested components', () => {
	const actual = shallow(<TableOfContents sections={sections} />);

	expect(actual).toMatchSnapshot();
});

it('should filter list when search field contains a query', () => {
	const searchTerm = 'but';
	const actual = shallow(<TableOfContents sections={[{ components }]} />);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('should filter section names', () => {
	const searchTerm = 'frm';
	const actual = shallow(<TableOfContents sections={sections} />);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('renderer should render table of contents', () => {
	const searchTerm = 'foo';
	const actual = shallow(
		<TableOfContentsRenderer
			classes={{}}
			items={<div>foo</div>}
			searchTerm={searchTerm}
			onSearchTermChange={noop}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should call a callback when input value changed', () => {
	const onSearchTermChange = jest.fn();
	const searchTerm = 'foo';
	const newSearchTerm = 'bar';
	const actual = shallow(
		<TableOfContentsRenderer
			classes={{}}
			items={<div>foo</div>}
			searchTerm={searchTerm}
			onSearchTermChange={onSearchTermChange}
		/>
	);

	actual.find('input').simulate('change', {
		target: {
			value: newSearchTerm,
		},
	});

	expect(onSearchTermChange).toBeCalledWith(newSearchTerm);
});
