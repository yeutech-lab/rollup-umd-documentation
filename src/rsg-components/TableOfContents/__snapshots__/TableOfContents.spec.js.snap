// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render table of contents 1`] = `
<div>
  <div>
    <div>
      <input
        aria-label="Filter by name"
        onChange={[Function]}
        placeholder="Filter by name"
        value="foo"
      />
    </div>
    <nav />
  </div>
</div>
`;

exports[`should filter list when search field contains a query 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Button",
          "sections": Array [],
          "slug": "button",
        },
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Input",
          "sections": Array [],
          "slug": "input",
        },
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Textarea",
          "sections": Array [],
          "slug": "textarea",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter list when search field contains a query 2`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm="but"
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Button",
          "sections": Array [],
          "slug": "button",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter section names 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Introduction",
          "sections": Array [],
          "slug": "introduction",
        },
        Object {
          "components": Array [
            Object {
              "name": "Button",
              "slug": "button",
            },
          ],
          "content": <ComponentsList
            items={
              Array [
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Button",
                  "slug": "button",
                },
              ]
            }
            useIsolatedLinks={false}
          />,
          "heading": true,
          "name": "Buttons",
          "sections": Array [],
          "slug": "buttons",
        },
        Object {
          "components": Array [
            Object {
              "name": "Input",
              "slug": "input",
            },
            Object {
              "name": "Textarea",
              "slug": "textarea",
            },
          ],
          "content": <ComponentsList
            items={
              Array [
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Input",
                  "slug": "input",
                },
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Textarea",
                  "slug": "textarea",
                },
              ]
            }
            useIsolatedLinks={false}
          />,
          "heading": true,
          "name": "Forms",
          "sections": Array [],
          "slug": "forms",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should filter section names 2`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm="frm"
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Forms",
          "sections": Array [],
          "slug": "forms",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should render a renderer 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Button",
          "sections": Array [],
          "slug": "button",
        },
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Input",
          "sections": Array [],
          "slug": "input",
        },
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Textarea",
          "sections": Array [],
          "slug": "textarea",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;

exports[`should render renderer with sections with nested components 1`] = `
<Styled(TableOfContents)
  onSearchTermChange={[Function]}
  searchTerm=""
>
  <ComponentsList
    items={
      Array [
        Object {
          "components": Array [],
          "content": false,
          "heading": false,
          "name": "Introduction",
          "sections": Array [],
          "slug": "introduction",
        },
        Object {
          "components": Array [
            Object {
              "name": "Button",
              "slug": "button",
            },
          ],
          "content": <ComponentsList
            items={
              Array [
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Button",
                  "slug": "button",
                },
              ]
            }
            useIsolatedLinks={false}
          />,
          "heading": true,
          "name": "Buttons",
          "sections": Array [],
          "slug": "buttons",
        },
        Object {
          "components": Array [
            Object {
              "name": "Input",
              "slug": "input",
            },
            Object {
              "name": "Textarea",
              "slug": "textarea",
            },
          ],
          "content": <ComponentsList
            items={
              Array [
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Input",
                  "slug": "input",
                },
                Object {
                  "content": false,
                  "heading": false,
                  "name": "Textarea",
                  "slug": "textarea",
                },
              ]
            }
            useIsolatedLinks={false}
          />,
          "heading": true,
          "name": "Forms",
          "sections": Array [],
          "slug": "forms",
        },
      ]
    }
    useIsolatedLinks={false}
  />
</Styled(TableOfContents)>
`;
